### DISCUSSION TITLE
SPECTER embeddings for textcat

### DISCUSSION THREAD

**Discussion by adamkgoldfarb** [explosion/spaCy#8884] [Help: Coding & Implementations] [answered]
Hello! 

I'm looking into using the [transformers implementation](https://huggingface.co/allenai/specter) of [AllenAI's SPECTER embeddings](https://github.com/allenai/specter) to feed a textcat_multilabel component (probably the ensemble). 

First, I'd like to confirm my understanding: SPECTER embeddings would not make sense as a drop-in replacement for a tok2vec component because they are document-level embeddings as opposed to token-level embeddings. Does that sound right?

Second, is there some facility for feeding the SPECTER embeddings as an additional feature for the textcat ensemble and/or as a replacement for the [bag-of-words component ](https://spacy.io/api/architectures#TextCatBOW)of the ensemble? In other words, does it make sense to try to use the SPECTER model as a drop-in replacement for the [model.linear_model]?

Thanks for any guidance or thoughts you might be able to offer!

**Comment by polm** [comment, marked as answer]
> First, I'd like to confirm my understanding: SPECTER embeddings would not make sense as a drop-in replacement for a tok2vec component because they are document-level embeddings as opposed to token-level embeddings. Does that sound right?

It's correct you can't replace the tok2vec like that. You might be able to use the Transformers wrapper - it's not clear to me how different your model is a from a typical Transformer- so you could look at that for reference.

> Second, is there some facility for feeding the SPECTER embeddings as an additional feature for the textcat ensemble and/or as a replacement for the bag-of-words component of the ensemble?

There isn't a good, easy way to pass extra features at the moment, see [the FAQ section on that][features]. I'm not sure the bag of words architecture in particular would be a good reference but using the textcat as a base and modifying it is definitely one way to do it.

[features]: https://github.com/explosion/spaCy/discussions/8226#:~:text=i%20want%20to%20add%20non-textual%20features

**Comment by adamkgoldfarb** [comment]
Thanks for the point re the FAQ! I had looked at that but wasn’t formulating the issue in my head as a non-textual one. I’ll read deep on those sections.

Thanks again for your help, always much appreciated.

Sent from a mobile device. Please excuse my brevity.
________________________________
